[FLINK-21504][checkpoint] Introduce notification of subsumed checkpoint

Myasuka opened the pull request:
## What is the purpose of the change
Introduce notification of subsumed checkpoint

## Brief change log

Introduce notification of subsumed checkpoint. Add a new `InternalCheckpointListener` and related implementations.

## Verifying this change

This change added tests and can be verified as follows:

SubtaskCheckpointCoordinatorTest

CheckpointCoordinatorTest


## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: no
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/StreamOperator.java:
```diff
@@ -43,7 +43,7 @@
  * @param <OUT> The output type of the operator
  */
 @PublicEvolving
-public interface StreamOperator<OUT> extends CheckpointListener, KeyContext, Serializable {
+public interface StreamOperator<OUT> extends InternalCheckpointListener, KeyContext, Serializable {
```
It seems wrong that a `@PublicEvolving` interface extends `@Internal` interface.
The root cause I think in the `StreamOperator` hierarchy, but that's a long story.

Semantically, I think operators should only be aware of checkpoint completion (to emit data for example); their backends should in addition know about the subsumption.

I see the following solutions to notify state backends about the subsumption
1. Modify `CheckpointListener` interface instead of introducing `InternalCheckpointListener` as you proposed initially
2. In `StreamOperatorWrapper`, get  `KeyedStateBackend` from the operator (if it's `AbstractStreamOperator`) and notify it directly (if it's `InternalCheckpointListener`)
3. Same, but via `StreamOperatorStateHandler`

WDYT?

cc: @pnowojski

pnowojski replied:
> It seems wrong to me that a @PublicEvolving interface extends @Internal interface.

+1 

What do you @rkhachatryan mean by:
> 1. Modify CheckpointListener interface instead of introducing InternalCheckpointListener as you proposed initially

?

Regardless I also think we shouldn't go through the operators if we don't want to make `notifyCheckpointSubsumed()` public.

rkhachatryan replied:
I meant to add a method `checkpointSubsumed()` to the `CheckpointListener` (or `lastSubsumedId` argument to its existing `notifyCheckpointComplete()` method).

pnowojski replied:
Do we want to expose `notifyCheckpointSubsumed()` on the DataStream API level in a `@Public` class?

Myasuka replied:
@pnowojski I think this needs more discussion here.
My previous idea was to add a default no-op in `CheckpointListener` so that could be public. Roman convinced me that this new interface does not make so much senses for most users, and could be acted as an internal API just like what `InternalKvState` did. 
However, current PR has to change several places to make the logic more clean, which makes me to think whether previous `AbstractStreamOperator` did not welcome such internal semantics here.
The main usage of `notifyCheckpointSubsumed` is for abstract keyed state backend, we can compare the relationship of `AbstractStreamOperator` with `AbstractKeyedStateBackend`. If `AbstractStreamOperator` is open for users to implement and `AbstractKeyedStateBackend` is actually also open for users to implement their own state backend.
If so, we might treat `notifyCheckpointSubsumed` as public also.
The only left questions are can we add a new method in `@Public` class and we need to remember to implement it in necessary places.
cc @rkhachatryan

pnowojski replied:
I'm also not sure if it makes sense to expose `notifyCheckpointSubsumed()` to the users. And if we don't have a good motivating example, I would prefer to not expose it. Why clog the public API at best? And at worst this could cause us problems down the road for providing backward compatibility/refactoring. If needed, we can always add this notification later. Removing and maintaining things is harder.

rkhachatryan commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractStreamOperator.java:
```diff
@@ -382,6 +382,11 @@ public void notifyCheckpointAborted(long checkpointId) throws Exception {
     }
 
+    @Override
+    public void notifyCheckpointSubsumed(long checkpointId) throws Exception {
+        stateHandler.notifyCheckpointSubsumed(checkpointId);
+    }
```
There is also `AbstractStreamOperatorV2`...

Myasuka replied:
I see, thanks for your reminder.

rkhachatryan reviewed (changes requested):
Thanks for the PR @Myasuka ,
I've left some comments, PTAL
(main concern is extending `InternalCheckpointListener` by `StreamOperator`).

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -1399,6 +1399,34 @@ public void notifyCheckpointAborted(final long checkpointID) {
         }
     }
 
+    @Override
+    public void notifyCheckpointSubsumed(long checkpointId) {
+        AbstractInvokable invokable = this.invokable;
+
+        if (executionState == ExecutionState.RUNNING && invokable != null) {
+
+            try {
+                invokable.notifyCheckpointSubsumedAsync(checkpointId);
+            } catch (RejectedExecutionException ex) {
+                // This may happen if the mailbox is closed. It means that the task is shutting
+                // down, so we just ignore it.
+                LOG.debug(
+                        "Notify checkpoint subsume {} for {} ({}) was rejected by the mailbox",
+                        checkpointId,
+                        taskNameWithSubtask,
+                        executionId);
+            } catch (Exception e) {
+                if (getExecutionState() == ExecutionState.RUNNING) {
+                    LOG.warn("Error while subsuming checkpoint {}.", checkpointId, e);
+                }
+            }
+        } else {
+            LOG.info(
+                    "Ignoring checkpoint subsume notification for non-running task {}.",
+                    taskNameWithSubtask);
+        }
+    }
+
```
Maybe try to de-duplicate the code? Existing `notifyCheckpointXXX` looks the same?

Myasuka replied:
Thanks for your suggestion, I will do the refactoring.

pnowojski commented on flink-core/src/main/java/org/apache/flink/api/common/state/InternalCheckpointListener.java:
```diff
@@ -0,0 +1,52 @@
+     * @throws Exception This method can propagate exceptions, which leads to a failure/recovery for
+     *     the task or job.
+     */
+    default void notifyCheckpointSubsumed(long checkpointId) throws Exception {}
```
Yes I second this comment.

Myasuka replied:
The original idea is to like current `CheckpointListener#notifyCheckpointAborted` which makes a default implementation to not let the normal users know. After making the interface not a default no-op implementation, I tried to make those abstract class implement this necessary interface to make the changes as least as possible.

pnowojski commented on flink-core/src/main/java/org/apache/flink/api/common/state/InternalCheckpointListener.java:
```diff
@@ -0,0 +1,52 @@
+ * for internal usage.
+ */
+@Internal
+public interface InternalCheckpointListener extends CheckpointListener {
```
Do we need this interface in the first place? The only value I can see of this is to have a common signature of this `notifyCheckpointSubsumed()`, which maybe is a good reason on it's own, which would make refactoring/renaming easier, but I'm not entirely sure...

Also I'm not entirely sure if `InternalCheckpointListener` should extend `CheckpointListener` or rather we should have another smaller interface `CheckpointSubsumedListener` that just adds this single method.

Myasuka replied:
My original idea is to add the new method in `CheckpointListener`. However, current `CheckpointListener` is already annotated as `public` and the semantics of a checkpoint subsumed should be acted as an internal operation only for runtime classes.
Maybe we can see the relationship as `InternalKvState` and `State`.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -138,7 +138,7 @@
                 TaskActions,
                 PartitionProducerStateProvider,
-                CheckpointListener {
+                InternalCheckpointListener {
```
+1

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1243,6 +1250,29 @@ private void completePendingCheckpoint(PendingCheckpoint pendingCheckpoint)
                         exception);
             }
+
+            try {
+                allPreviousCheckpoints.removeAll(completedCheckpointStore.getAllCheckpoints());
+                if (!allPreviousCheckpoints.isEmpty()) {
```
The original idea is to not change the API, and I think a proper place is to change the return result of `CompletedCheckpointStore#addCheckpoint` from `void` to the subsumed `CompletedCheckpoint`.

rkhachatryan replied:
I think returning **subsumed** checkpoint from `addCheckpoint` is not very straightforward (as a caller, I'd expect **added** checkpoint to be returned).
I'd rather add `long getLatestSubsumedCheckpointID();` method or something like that.

Myasuka replied:
Actually, current `CheckpointStore#addCheckpoint` has the meaning that this is an atomic operation which will add the completed checkpoint and subsumed an older checkpoint.` I think return the subsumed checkpoint is clearly for this operation.
If we adopt the API `long getLatestSubsumedCheckpointID()`, we will miss some information that when the checkpoint is subsumed and we might also need to check whether the latest subsumed checkpoint ID has been changed.

rkhachatryan replied:
> Actually, current CheckpointStore#addCheckpoint has the meaning that this is an atomic operation which will add the completed checkpoint and subsumed an older checkpoint. I think return the subsumed checkpoint is clearly for this operation.

To me, it's only clear that it might subsume some older checkpoints. It is unclear whether returned is the first subsumed, the latest subsumed, or the added one.

> If we adopt the API long getLatestSubsumedCheckpointID()`, we will miss some information that when the checkpoint is subsumed and we might also need to check whether the latest subsumed checkpoint ID has been changed.

I think there is no requirement for the CheckpointCoordinator to report strictly latest ID (it will be outdated right after the RPC). It must only be non-decreasing. If it somehow becomes greater than the added one we could use the smallest from them.

And currently Store is used from the same thread, so calling `addCheckpoint` and then `getLatestSubsumed` is safe (I don't think this concurrency model should be changed).

Myasuka replied:
I think we can refer to the documentation of `CompletedCheckpointStore#addCheckpoint`:
~~~ java
Only a bounded number of checkpoints is kept. When exceeding the maximum number of
retained checkpoints, the oldest one will be discarded.
~~~
And all current implementations obey this role to make this method atomic and discard the oldest checkpoint internally. That's why, to me, changing this method to return the oldest discarded checkpoint looks so straightforward.

rkhachatryan replied:
Yes, it's clear when one goes through the documentation. 
However, I think it's better to make the API clear even without the need to read the javadoc.

Myasuka replied:
I see, maybe the previous name of `addCheckpoint` is not clear for what it really does. It actually does `addCheckpointAndSubsumeOldestOne`.

rkhachatryan replied:
The new name doesn't seem to reflect what the method actually does.
It doesn't necessarily subsume **one** checkpoint (or **any** checkpoints at all).

The correct name would be `addAndMaybeSubsumeOldestAndReturnTheLatestSubsumedIfAny` - which clearly indicates to me that the new functionality should be in a separate method.

Myasuka replied:
Yes, I agree that the returned subsumed checkpoint might be null, but I can't see why we need to add a separate method? If so, who will call the original `addCheckpoint` method?

rkhachatryan replied:
> Yes, I agree that the returned subsumed checkpoint might be null

From the (current) name, it's unclear which checkpoint is returned, even if not null.

> I can't see why we need to add a separate method? 
> who will call the original addCheckpoint method?

In my opinion, it would make a more clear API.
`CheckpointCoordinator` will call it (same as now). Let me demonstrate this: 
https://github.com/rkhachatryan/flink/commit/10242b7aef2f4f2db94c924f940928fce4ab0fc9#diff-2c7004a2d412c3566de5ff6fb9e6d027742328c2acad60685e47ce4ba9df0810R1286

WDYT?

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/state/TaskStateManager.java:
```diff
@@ -42,7 +42,7 @@
  * state of operator instances in the task for restore purposes.
  */
-public interface TaskStateManager extends CheckpointListener, AutoCloseable {
+public interface TaskStateManager extends InternalCheckpointListener, AutoCloseable {
```
I previously thought that `TaskStateManagerImpl` could also leverage the `notifyCheckpointSubsumed` API to help remove local states. This is actually another story and not seems so valuable in current implementation.
I will remove this in next commit.

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/api/operators/AbstractUdfStreamOperator.java:
```diff
@@ -136,6 +136,11 @@ public void notifyCheckpointAborted(long checkpointId) throws Exception {
         }
     }
 
+    @Override
+    public void notifyCheckpointSubsumed(long checkpointId) throws Exception {
+        super.notifyCheckpointSubsumed(checkpointId);
+    }
```
Hmm, my fault. Since current `userFunction` is not instance of `InternalCheckpointListener`, we could remove this.

Myasuka commented:
@rkhachatryan @pnowojski , thanks for your review. I have refactored this PR with two commits:
1beb902: Refactor `CompletedCheckpointStore#addCheckpoint` to `CompletedCheckpoint#addCheckpointAndSubsumeOldestOne`.

36cfdc4: Introduce the notification of subsumed checkpoint which addressed your comments.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointSubsumeHelper.java:
```diff
@@ -67,6 +75,7 @@ public static void subsume(
             }
             // Don't break out from the loop to subsume intermediate savepoints
         }
+        return lastSubsumedCheckpoint;
```
I think it would be more clear if the latest subsumed checkpoint is recorded by the store in it's subsume action (because of the same ambiguity about what is returned from this method).

Myasuka replied:
Maybe we could record the latest subsumed checkpoint in the store, however, this might not get the core idea of current `CompletedCheckpointStore#addCheckpoint`. 
You can take a look at `DefaultCompletedCheckpointStore#addCheckpoint`, once the `checkpointStateHandleStore.addAndLock(path, checkpoint)` failed, it would trigger the subsume process anymore to ensure strict high-availability. The timing to subsume checkpoint is very important and just record a subsumed checkpoint but not known when that one is subsumed would lose information.

rkhachatryan replied:
WDYT about [this](https://github.com/rkhachatryan/flink/commit/10242b7aef2f4f2db94c924f940928fce4ab0fc9#diff-1bc8de30a2deacb9128e6bb8c13169b3cb692beddb08e6724f1cf2b550435effR264) approach?

Myasuka replied:
Thanks for Roman sharing the implementation of adding another method `CompletedCheckpointStore#getLatestSubsumedCheckpointID`. However, I still prefer to leverage previous `CompletedCheckpointStore#addCheckpoint` to know the last subsumed checkpoint for reasons:

1. In general, we should make the minimal changes to existing interfaces.
2. Previous method of `CompletedCheckpointStore#addCheckpoint` already act the role to subsume checkpoint/savepoint when adding a new complete checkpoint. We could achieve our goal If we just returned the subsumed checkpoint in previous method instead of return nothing.
3. Adding method of `CompletedCheckpointStore#getLatestSubsumedCheckpointID` would introduce another method and would also make us lost the timing when the checkpoint is subsumed.

After offline discussion with each other, we think maybe involving more guys could help this discussion more clear. Could you share some thoughts here? @StephanEwen @pnowojski

pnowojski replied:
edit: actually after removing field `@Nullable private CompletedCheckpoint lastSubsumedCheckpoint;` I don't see actually much difference between those two approaches. Is there a problem that I'm missing? 

Note it doesn't help that the changes are not squashed, ant that the counter proposal is also a separate commit on top of this PR. It makes reviewing and comparing both options very difficult.

Myasuka replied:
@pnowojski , thanks for your review! Sorry for late reply due to recent too busy internal work.

The reason I prefer the solution to return the checkpoint to subsume is that we might not susbume checkpoint each time especially when we adding a complete savepoint (see [FLINK-21351](https://issues.apache.org/jira/browse/FLINK-21351)). Just use a getter would lost such information.

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CompletedCheckpointStore.java:
```diff
@@ -45,8 +47,12 @@
      */
-    void addCheckpoint(
+    @Nullable
+    CompletedCheckpoint addCheckpointAndSubsumeOldestOne(
```
The new name doesn't seem to reflect what the method actually does.
It doesn't necessarily subsume **one** checkpoint (or **any** checkpoints at all).

The correct name would be `addAndMaybeSubsumeOldestAndReturnTheLatestSubsumedIfAny` - which clearly indicates to me that the new functionality should be in a separate method.

rkhachatryan reviewed (changes requested):
Thanks for updating the PR @Myasuka 
I have some further comments, PTAL.

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -229,6 +229,8 @@
 
     private final ExecutionAttemptMappingProvider attemptMappingProvider;
 
+    @Nullable private CompletedCheckpoint lastSubsumedCheckpoint;
+
```
This is just for easy of testing in `CheckpointCoordinatorTest`.

rkhachatryan replied:
I think the added complexity in `CheckpointCoordinator` doesn't worth it as it's a critical component.
And it's more valuable to test the actual functionality - i.e. what ID is sent in the notifications.

Myasuka replied:
Thanks for your suggestion, I will consider to change this.

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointCoordinator.java:
```diff
@@ -1220,8 +1224,15 @@ private void completePendingCheckpoint(PendingCheckpoint pendingCheckpoint)
+                                completedCheckpoint,
+                                checkpointsCleaner,
+                                this::scheduleTriggerRequest);
+                if (lastSubsumed != null && lastSubsumed.discardOnSubsume()) {
```
I think this needs more discussion here. After we store savepoint in checkpoint store, the actual meaning of `CompletedCheckpointStore#addCheckpoint` has changed: we might subsume a savepoint and a checkpoint in the same turn when adding a new checkpoint. My understanding is that we would not need to notify tasks of subsumed savepoint as it should not be owned on task side.

rkhachatryan replied:
I think we shouldn't make such a distinction.
If TM needs to distinguish for some particular task (state management), it can be done on TM side.
One practical reason to notify TMs is that subsuming a savepoint allows them subsume older checkpoints.
A more general reason that the community I think is moving towards miminizing the difference in handling of savepoints and checkpoints.

Myasuka replied:
If we also send notification of subsumed savepoint, there would be a problem that we might subsume a savepoint and a checkpoint in the same turn when adding a new checkpoint. For this case, which checkpoint id should we notify?

I see your PR of https://github.com/apache/flink/pull/16575 also consider the savepoint case on TM side, was that the reason you prefer to send notification of savepoints?

rkhachatryan replied:
> which checkpoint id should we notify?

I think it should always be the latest (regardless whether it's savepoint or checkpoint).

> I see your PR of #16575 also consider the savepoint case on TM side, was that the reason you prefer to send notification of savepoints?

No, I mentioned the reasons above (the main one is generality). 

Regarding the PR (#16575), TM needs to distinguish when discarding the state anyways. This is because it may have non-confirmed savepoints (that might be aborted); at the same, a newer checkpoint might be subsumed; which would mean that those older savepoints are subsumed.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -1399,6 +1399,33 @@ public void notifyCheckpointAborted(final long checkpointID) {
         }
     }
 
+    public void notifyCheckpointSubsumed(long checkpointId) {
+        AbstractInvokable invokable = this.invokable;
+
+        if (executionState == ExecutionState.RUNNING && invokable != null) {
+
+            try {
+                invokable.notifyCheckpointSubsumedAsync(checkpointId);
+            } catch (RejectedExecutionException ex) {
+                // This may happen if the mailbox is closed. It means that the task is shutting
+                // down, so we just ignore it.
+                LOG.debug(
+                        "Notify checkpoint subsume {} for {} ({}) was rejected by the mailbox",
+                        checkpointId,
+                        taskNameWithSubtask,
+                        executionId);
+            } catch (Exception e) {
+                if (getExecutionState() == ExecutionState.RUNNING) {
+                    LOG.warn("Error while subsuming checkpoint {}.", checkpointId, e);
+                }
+            }
+        } else {
+            LOG.info(
+                    "Ignoring checkpoint subsume notification for non-running task {}.",
+                    taskNameWithSubtask);
+        }
+    }
```
You have resolved my previous comment about code duplication, but it still looks very similar as `notifyCheckpointAborted()` and `notifyCheckpointCompleted()`?

Myasuka replied:
I mistakenly thought that you left the comments on `SubtaskCheckpointCoordinatorImpl` class, and I refactored that class to remove code duplication. And after your remider, I see your comments are under `Task` class, and I will refactor later.

Myasuka commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/StreamOperatorWrapper.java:
```diff
@@ -100,6 +102,16 @@ public void notifyCheckpointComplete(long checkpointId) throws Exception {
 
+    public void notifyCheckpointSubsumed(long checkpointId) throws Exception {
+        if (!closed) {
+            if (wrapped instanceof AbstractStreamOperator) {
+                ((AbstractStreamOperator<?>) wrapped).notifyCheckpointSubsumed(checkpointId);
+            } else if (wrapped instanceof AbstractStreamOperatorV2) {
+                ((AbstractStreamOperatorV2<?>) wrapped).notifyCheckpointSubsumed(checkpointId);
+            }
```
Adding the method was to just follow operator hierarchy and we could extend the usage to other componments ont only keyed state backends if necessary in the future.
If you don't like to add methods in `PublicEvolving` class, I think it's fine and I could revert to call `getKeyedStateBackend()`.

pnowojski replied:
That's a tough call. It should follow the hierarchy and `StreamOperator` shouldn't be the public API (we are planning to remove it from the public api at some point). But as it is, it doesn't matter if you add this method to `StreamOperator` or `AbstractStreamOperator`, because all practical cases for implementing a custom `StreamOperator` are always extending from `AbstractStreamOperator`. So I would propose to call `notifyCheckpointSubsumed()` via `getKeyedStateBackend()` for now. Later, once we remove `StreamOperator` from public API we can change it back.

Myasuka replied:
Got it, thanks for your explanation.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -339,78 +343,107 @@ public void checkpointState(
+                            new CancellationException("checkpoint aborted via notification"),
+                            false);
+                },
+                (opw) -> opw.getStreamOperator().notifyCheckpointAborted(checkpointId),
+                (tsm) -> tsm.notifyCheckpointAborted(checkpointId));
```
please no abbreviations. By Looking at this code I have no idea what `opw` or `tsm` are.

pnowojski commented on flink-streaming-java/src/main/java/org/apache/flink/streaming/runtime/tasks/SubtaskCheckpointCoordinatorImpl.java:
```diff
@@ -339,78 +343,107 @@ public void checkpointState(
+                null);
+    }
+
+    private void notifyCheckpoint(
+            long checkpointId,
+            OperatorChain<?, ?> operatorChain,
+            Supplier<Boolean> isRunning,
+            String logFormatIfNotRunning,
+            String logFormatIfFinishedOnRestore,
+            String logFormatIfNotificationStart,
+            @Nullable Runnable runnableBeforeNotifyOperator,
+            @Nonnull ThrowingConsumer<StreamOperatorWrapper<?, ?>, Exception> operatorConsumer,
+            @Nullable ThrowingConsumer<TaskStateManager, Exception> taskStateManagerConsumer)
+            throws Exception {
```
Instead of using `Runnable` and `ThrowingConsumer` in order to deduplicate the code, I would suggest to consider deduplicating it by decomposing this `notifyCheckpoint` method into smaller parts. For example:

```

public void notifyCheckpointAborted(...) {
  if (notifyCheckpointOperationrShouldBeExecuted()) {
    // inlined content of the `runnableBeforeNotifyOperator`
    
    operatorChain.forEach((operatorWrapper) -> operatorWrapper.notifyCheckpointSubsumed(checkpointId));
    
    env.getTaskStateManager().notifyCheckpointAborted(checkpointId);
  }
}

public void notifyCheckpointSubsumed(...) {
  if (notifyCheckpointOperationrShouldBeExecuted()) {
    operatorChain.forEach((operatorWrapper) -> operatorWrapper.notifyCheckpointAborted(checkpointId));
  }
}

private boolean notifyCheckpointOperationrShouldBeExecuted(...) {
  // isRunning/isFinishedOnRestore checks + logging
}
```

Yes, we would still have a bit of code duplication, but it would avoid the problem of generic `Runnable` `ThrowingConsumer` that are mangling/hiding what the code is doing. In your current version to understand what this code is doing, you have to keep jumping back and forth between `notifyCheckpointAborted()` method (where those `Runnable` and `ThrowingConsumers` are defined) and the actual `notifyCheckpoint` implementation. What makes it even worse, is that you can not use "find implementations" from the IDE to help you, as there are just so many `Runnable`s in the project. So maybe a bit of code duplication is not that bad in that case? (I'm not 100% sure)

Myasuka replied:
The refactor also take the `enum` of `NotifyCheckpointOperation` as `Task` did.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -1352,84 +1352,73 @@ public void triggerCheckpointBarrier(
     }
 
-    public void notifyCheckpointSubsumed(long checkpointId) {
+    private void notifyCheckpoint(
+            long checkpointId,
+            Consumer<AbstractInvokable> invokableNotifier,
+            String logFormatIfRejected,
+            String logFormatIfFailedToNotify,
+            @Nullable Consumer<Throwable> handleThrowableIfTaskRunning,
+            String logFormatIfNotRunning) {
```
For me this method has the same problem as in my previous comment with generic Consumer's. However here I would propose to add an `enum NotifyCheckpointOperation` with three values `ABORT`, `COMPLETE`, `SUBSUME`, and just switch/case it to replace `Consumer<AbstractInvokable> invokableNotifier`.

On the other hand `handleThrowableIfTaskRunning` could be replaced by boolean flag `boolean shouldFailExternallyOnException`, so the signature of this method could be:
```
    private void notifyCheckpoint(
            long checkpointId,
            NotifyCheckpointOperation notifyCheckpointOperation,
            boolean shouldFailExternallyOnException)
```

Thirdly, the log/error formats I think could be handled also easier, by just using something like:
```
String.format("Notify checkpoint {} {} for {} ({}) was rejected by the mailbox", notifyCheckpointOperation.toString(), ...);
```
instead of
```
String.format("Notify checkpoint abort {} for {} ({}) was rejected by the mailbox",  ...);
```

Myasuka replied:
Thanks for the suggestion of introducing `enum` of `NotifyCheckpointOperation`, I think this could help reduce code duplication and also make the code more readable.

Myasuka commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointSubsumeHelper.java:
```diff
@@ -46,18 +48,24 @@
             throws Exception {
         if (checkpoints.isEmpty() || checkpoints.size() <= numRetain) {
-            return;
+            return null;
```
I see, I'll refactor API to use `Optional`

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointSubsumeHelper.java:
```diff
@@ -46,18 +46,24 @@
+            return Optional.empty();
         }
         CompletedCheckpoint latest = checkpoints.peekLast();
+        CompletedCheckpoint lastSubsumedCheckpoint = null;
```
optional nit: `Optional<CompletedCheckpoint> lastSubsumedCheckpoint = Optional.empty();`?

Myasuka replied:
Thanks for the suggestion, I will refactor it in the next upadte.

pnowojski commented on flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java:
```diff
@@ -1405,59 +1406,80 @@ private void declineCheckpoint(
+                                            t));
+                        }
+                        break;
+                    case SUBSUME:
+                        // just rethrow the throwable out as we do not expect notification of
+                        // subsume could fail the task.
+                        ExceptionUtils.rethrow(t);
```
Why do we need a special handling here?

What do you mean "We do not expect notification could fail the task"? Do you mean that this error should never happen? If so, I would keep the error handling unified regardless.

Myasuka replied:
From my point of view, we need to ensure task failover directly on the case of checkpoint complete notification to avoid unexpected data lost. However, I don't think we should also apply this rule on checkpoint subsumed notification as there is no strict data completeness request here. WDYT?

pnowojski replied:
I'm a little bit thorn apart here.

Wouldn't the same logic apply to `notifyCheckpointAbort()`? 

On the other hand, even in the case of an exception during `notifyCheckpointComplete()`. As long as checkpoint is failed, there won't be any data loss. Data will be committed in the next checkpoint OR if there are enough checkpoint failures, the whole job will failover because of the exceeded number of tolerable checkpoint failures.

I think it could be argued both directions for either of those methods (abort/complete/subsume), so maybe for the sake of simplicity and consistency it would be best to keep them behaving in the same way?

Myasuka replied:
I see, for the simplicity I will refactor this to keep the same behavior.

pnowojski replied:
Now I remembered some past discussion about the behaviour on failing tasks or not during checkpointing. Our previous conclusion was indeed no to fail the job if not needed. We have to fail the job, if the exception happened in for example the sync phase of snapshotting, as otherwise operators/statebackends/user functions might end up in an inconsistent state. However failures from for example async phase can be handled by the `CheckpointCoordinator` and `CheckpointFailureManager`.

`notifyCheckpointComplete` and `notifyCheckpointAborted` has to fail the job, at least in some/most cases, because if the failure has originated in the user code, that user code might be in an invalid state. For example:
```
void notifyCheckpointComplete() {
  a++;
  maybeThrowException();
  b++;
}
```
after throwing `a` and `b` will be inconsistent. 

As we don't expose `notifyCheckpointSubsumed()` to the users, this argument doesn't hold here. So the only question is our internal data structures are expected to keep working just fine if an exception indeed happens in this call?

Myasuka replied:
From my point of view, our internal data structures are expected to keep working fine if come across an exception here. Some customized keyed state backend might introduce IO operations within `notifyCheckpointSubsumed()`, and the task should not fail as those IO operations are best-effort-to-do. That's why my implementation would not fail here.

pnowojski replied:
👍 
So let's keep it as it is. (failure of `notifyCheckpointSubsumed` shouldn't cause the task to failover automatically).

pnowojski reviewed (commented):
Thanks for the update @Myasuka , I've left a couple of comments.

pnowojski reviewed (approved):
I think LGTM assuming green azure and that @rkhachatryan is also fine with the change.

rkhachatryan reviewed (approved):
I don't have any further comments.

Myasuka commented:
My local CI turns green: https://myasuka.visualstudio.com/flink/_build/results?buildId=362&view=results

rkhachatryan commented on flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/DefaultCompletedCheckpointStore.java:
```diff
@@ -133,17 +134,23 @@ public void addCheckpoint(
+                                        checkpointsCleaner,
+                                        postCleanup));
+
+        if (subsume.isPresent()) {
+            LOG.debug("Added {} to {} without any older checkpoint to subsume.", checkpoint, path);
+        } else {
+            LOG.debug("Added {} to {} and subsume {}.", checkpoint, path, subsume);
+        }
```
Shouldn't it be the opposite?

Myasuka replied:
Yes, I think the log info should swap the position. Since these logs have been removed after refactor, I think we don't need to create a hotfix now.

rkhachatryan replied:
That's true but the code was refactored in 1.16. 
It's still swapped in 1.15 so I created FLINK-29158 to fix it.
